### DISCUSSION TITLE
Netbox tenant to tenant migration

### DISCUSSION THREAD

**Discussion by F3d0s** [netbox-community/netbox#12088] [Help Wanted!] [answered]
As we got a new company in the group, we would like to migrate their Netbox tenant database to our current Netbox. Bot run version 3.4.4.

I've tried exporting `csv` or `YAML` files of every object type and importing them to our Netbox but unfortunately, very few have gone well, some had an issue with the `csv` file which I had to modify (i.e. first line all lowercase and no spaces), or other issues. This procedure seems to be very slow and I don't have all this time.

Is there a way to do a tenant-to-tenant migration without exporting/importing manually every section and dealing with many issues?

**Comment by candlerb** [comment]
It sounds like you're asking "can I export a subset of data from one Netbox and import it into another"? I'm afraid this is very difficult, as you've already found, and is not really a supported use case.

If the target is a fresh Netbox, your simplest approach may be to clone the whole database, then delete everything else that's not needed in the target instance.

I have done migrations via SQL-level export and import in the past (SELECT and INSERT), but if you're importing into a Netbox instance which  has existing data, you would have to modify the exported data to replace any primary keys which clashed with existing values - and find and change all the references to those primary keys - prior to importing.  You'd also have to update the sequences to skip past the new values assigned.

The other options are also very difficult to do a high-fidelity export and import:
* The CSV export and import formats are different, and don't include the full set of options (e.g. multi-terminations for cables)
* There are circular dependencies in foreign keys, e.g. Device -> Primary IP requires Device <- Interface <- IPAddress to be set up first, and you can't create objects of different types in the same transaction via either REST API or CSV import. So you'd have to import the device without Primary IP, import the interfaces, import the IP addresses, and finally patch up the Primary IP.
* Objects in Netbox don't have UUIDs, only the sequentially-generated integer database IDs. If you create an object on the target system via CSV or REST then it will have a different ID to the source system, and you have to keep track of these mappings externally.

**Reply by F3d0s** [reply]
> The CSV export and import formats are different and don't include the full set of options (e.g. multi-terminations for cables)

I wonder why there is even the option to export the data if the format is not even compatible with Netbox itself...

**Reply by jeremystretch** [reply]
https://github.com/netbox-community/netbox/wiki/Frequently-Asked-Questions#why-cant-i-pass-exported-data-directly-into-the-import-form

**Comment by jeremystretch** [comment]
> Is there a way to do a tenant-to-tenant migration without exporting/importing manually every section and dealing with many issues?

Yes, you can easily script this out using the [pynetbox](https://github.com/netbox-community/pynetbox) API client. The docs list a recommended [order of operations](https://docs.netbox.dev/en/stable/getting-started/planning/#order-of-operations) for populating new data.

However, the much bigger challenge is in validating the data prior to replication: identifying duplicate or invalid objects, ensuring consistency with existing conventions (tagging, etc.), determining which custom fields need to be replicated, and so forth. Make sure you have a written plan in plan to address these questions before beginning, or you'll end up making more work for yourself.

**Comment by F3d0s** [comment, marked as answer]
In the end, I did this by exporting the other Netbox data's csv files, modifying them and importing them to my Netbox installation. It was a tedious and long process especially because I didn't know how to modify the header and some data but in the end, I could do it.

It would be great if there was a migration tool for operations like that.
